client/src/language/codeActionProvider.ts

```diff
@@ -0,0 +1,101 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import vscode from 'vscode'
+import { requestsManager } from './RequestManager'
+import { embeddedLanguageDocsManager } from './EmbeddedLanguageDocsManager'
+import { getEmbeddedLanguageDocRange, getIndentationOnLine, getOriginalDocPosition } from './utils'
+import { logger } from '../lib/src/utils/OutputLogger'
+import { type Range } from 'vscode-languageclient'
+
+export class BitbakeCodeActionProvider implements vscode.CodeActionProvider {
+  async provideCodeActions (document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<vscode.CodeAction[]> {
+    const diagnostics = context.diagnostics
+    const actions = await Promise.all(
+      diagnostics.map(async (diagnostic) => await buildActionFromDiagnostic(document, diagnostic))
+    ).then((results) => results.flat())
+    return actions
+  }
+}
+
+const buildActionFromDiagnostic = async (document: vscode.TextDocument, diagnostic: vscode.Diagnostic): Promise<vscode.CodeAction[]> => {
+  const originalRange = diagnostic.range
+  const embeddedLanguageType = await requestsManager.getEmbeddedLanguageTypeOnPosition(document.uri.toString(), originalRange.start)
+  if (embeddedLanguageType === undefined || embeddedLanguageType === null) {
+    return []
+  }
+  const embeddedLanguageDocInfos = embeddedLanguageDocsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), embeddedLanguageType)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return []
+  }
+  const embeddedLanguageTextDocument = await vscode.workspace.openTextDocument(embeddedLanguageDocInfos.uri)
+  const embeddedRange = getEmbeddedLanguageDocRange(
+    document,
+    embeddedLanguageTextDocument,
+    embeddedLanguageDocInfos.characterIndexes,
+    originalRange
+  )
+
+  const tempActions = await vscode.commands.executeCommand<vscode.CodeAction[]>(
+    'vscode.executeCodeActionProvider',
```

Comment:
It pulls the code actions for the given file on the given range. For example, if you pass to the command a Python file and a range, you will get the same array of code actions that you would see if you open the Python file manually with VS Code and put your cursor within the given range.

It works for any language. For example, I use the same command in the tests to pull the code actions for BitBake, which are provided by our extension. It would also work on Bash with the code actions provided by Shellcheck, which we haven't added yet to our dependencies.

Reply:
Do you have example syntax errors in bash functions for which shellcheck provides code action?
Would be nice to test one as well in the integration test, even just it's existent.

Reply:
I've seen only two types of code actions for shellcheck. It either proposes to see the explanation for the error into the wiki, or to disable the error. I don't think people want to add `#shellcheck disable=` into their code, but they might be interested by the quick link to see the explanation from the wiki.